### PR TITLE
Check bank capitalization

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

- No check for bank capitalization.

- Also, the current offline bank capitalization check doesn't catch overflows.

#### Summary of Changes

On accounts hashing, load the total accounts capitalization and compare with bank.capitalization.

Fixes #

#### context

reboot of https://github.com/solana-labs/solana/pull/10937

### CODE REVIEW

**Comment by ryoqun** [comment]
btw, because the testnet ledger was restarted, it's now safe to start to enforce this everytime validator starts. (thanks @mvines).

File: runtime/src/bank.rs
Comment (ryoqun) [inline runtime/src/bank.rs:2846]: This has been solely used for the `ledger-tool capitalization` subcommand.

File: runtime/src/accounts_db.rs
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1577]: This `executable` is needed to be unblocked from #11928

Comment (ryoqun) [reply runtime/src/accounts_db.rs:1577]: @sakridge  sorry for rather intrusive revert of your changes (65342994f05b309b92b137533dbc48f83f905708). But, I think this account.owner-based condition is just enough for capitalization calculation purpose, instead of carrying around `pubkeys`. That's because I want to avoid adding bank fields unless really needed. And new check is state-less, so less risk of introducing bugs. And lastly, this will be faster or equivalent in terms of performance, compared to 1 lookup of `HashSet` for each account.

Was there any concern I'm missing for choosing using `builtin_programs` and `genesis_cap_exempt_ids`?.

As for deprecated storage accounts/rewards pool, I'm going to address them first. It's living only in testnet at this moment.

File: runtime/src/accounts_db.rs
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1555]: I have yet to check perf comparison of checked u64 addition vs unchecked u128 addition....

Also, consider par_iter() + fold()?

**Comment by aeyakovenko** [comment]
@ryoqun Let's say we had an inflation bug, like an overflow in the contract, would this catch it?  I want to make sure that we can compute the max possible supply in the snapshot and fail if it exceeds it.

File: runtime/src/accounts_db.rs
Comment (sakridge) [reply runtime/src/accounts_db.rs:1577]: No, that's great. Definitely better to avoid adding bank fields if we can get away with it. I wasn't actually sure about the exact conditions where capitalization wasn't applied.

File: runtime/src/bank.rs
Comment (ryoqun) [inline runtime/src/bank.rs:2861]: https://github.com/solana-labs/solana/pull/11927#issuecomment-684114178

> btw, because the testnet ledger was restarted, it's now safe to start to enforce this everytime validator starts. (thanks @mvines).

very sadly, `sum()` was so subtle. my checked math in-brain type checker got tricked. ;) **This does with unchecked math.** So, testnet capitalization was still off-calcualtion combined with pre-mature removal of storage account special casing at #9992.

I'll first have to create another gated pr to really remove rewards_pool in testnet.

File: runtime/src/accounts_db.rs
Comment (sakridge) [reply runtime/src/accounts_db.rs:1577]: > And lastly, this will be faster or equivalent in terms of performance, compared to 1 lookup of HashSet for each account.

Might be debatable, because this sysvar::check_id has to do a compare with all existing sysvars. The hashset just has to hash the account value and then check the table to see if it exists in the set. It would probably touch a lot less memory.

File: ledger/src/blockstore_processor.rs
Comment (ryoqun) [inline ledger/src/blockstore_processor.rs:486]: Without this check, validators/cluster might continue to create newer blocks combined with automatic restart practiced from the ops perspectives. That's because every restarts delays next snapshot slot by the snapshot interval (usually 100). ;)

Comment (ryoqun) [reply ledger/src/blockstore_processor.rs:486]: Well, I'll do write this in-souce comment as well.

File: runtime/src/accounts_db.rs
Comment (sakridge) [reply runtime/src/accounts_db.rs:1577]: I see. That `calculate_capitilzation` function didn't exist when I started this PR and I missed that it was added.

File: runtime/src/accounts_db.rs
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1627]: Well, I wanted AtomicU128, but it looks like this still isn't stable: https://github.com/rust-lang/rust/issues/32976

**Comment by ryoqun** [comment]
> @ryoqun Let's say we had an inflation bug, like an overflow in the contract, would this catch it? I want to make sure that we can compute the max possible supply in the snapshot and fail if it exceeds it.

@aeyakovenko Yeah, I've tested this locally. These catches them as the last protection mechanism.

For both inflation and overflow in contracts, these will be protected by their specific protections, btw. So, I've manually disabled them while testing locally. I've added some sanity assertion: https://github.com/solana-labs/solana/blob/11ac4eb21d8cefb194d0813a180240e6d89780cf/runtime/src/bank.rs#L932-L940 for inflation and this is the protection for contracts: https://github.com/solana-labs/solana/blob/11ac4eb21d8cefb194d0813a180240e6d89780cf/runtime/src/message_processor.rs#L559-L578

File: runtime/src/accounts_db.rs
Comment (ryoqun) [reply runtime/src/accounts_db.rs:1577]: > Might be debatable, because this sysvar::check_id has to do a compare with all existing sysvars.

Fortunately, this is not true. We're comparing `account.owner` not indivisual actual sysvar addresses. Maybe you're referring to this: https://github.com/solana-labs/solana/blob/8bcc8edf47b556c7177bb9058dacc62dd101c575/sdk/src/sysvar/mod.rs#L21-L31. But we're intead doing this: https://github.com/solana-labs/solana/blob/8bcc8edf47b556c7177bb9058dacc62dd101c575/sdk/macro/src/lib.rs#L48-L50

So, simple byte-wite comparison of `Pubkey` (note: this is `derive(Copy)`).

And, because almost all  of comparison is negative, this is super friendly for branch prediction, bailing out after few prefix bytes. And the required data is a single `Pubkey` or two if you include `NativeLoader`, so this will be cpu-cache super friendly as well. :)

At least, HashSet requires hashing the whole Pubey, without any short-circuitting. So, I think this will be faster in theory. I haven't measured it, though... ;)

Comment (sakridge) [reply runtime/src/accounts_db.rs:1577]: Yea, I mean you have to do a load for each of those sysvar Pubkey which even if you only read the 1st byte, you still do a load on a fresh cache line, and each pubkey will likely be on different cache lines. So you have to 10 loads even if its' just looking at the first byte of the pubkey.

To do the hash of the whole pubkey means you only need to load in a single 64-byte cache line since the pubkey will be contiguous in memory. The arithmetic instructions to do the hash are then essentially free. Then you need 1 more load at the HashSet table offset.

So 2 loads vs. 10.

File: runtime/src/accounts_db.rs
Comment (ryoqun) [reply runtime/src/accounts_db.rs:1627]: done https://github.com/solana-labs/solana/pull/11927/commits/51d532f0b1f8e871f1ef3cf7bcf159701665515a

File: runtime/src/accounts_db.rs
Comment (ryoqun) [reply runtime/src/accounts_db.rs:1577]: > > HashSet requires hashing the whole Pubey, without any short-circuitting.

> The arithmetic instructions to do the hash are then essentially free.

Oh, SipHash is that pipeline-freindly and fast? I never knew..

And, thanks for explaining.

So just to be sure, the 10 loads is for `is_sysvar_id`, which we won't use here.

We're going to use `check_id`, which should cost 2 loads according to your explanation (account's `owner` pubkey itself and the literal `Sysvar1111111111111111111111111111111111111`).

File: runtime/src/bank.rs
Comment (ryoqun) [reply runtime/src/bank.rs:2861]: > I'll first have to create another gated pr to really remove rewards_pool in testnet.

This is done #12011

**Comment by ryoqun** [comment]
@sakridge Could you review this? I have still something to do (add test for overflow; benchmark checked u64 vs u128 for summing) but mostly this should be done.

What I have added to your original pr:
- Really Inhibit automatic at blockstore_processor restart after capitalization mismatch is detected (otherwise, I observed the cluster would manage to progress while being tripped at every snapshot interval...)
- Don't miss overflowing capitalization (so that I tweaked the data passing for the verification loops).

File: runtime/src/accounts_db.rs
Comment (ryoqun) [reply runtime/src/accounts_db.rs:1555]: Well, I played with `godbolt` a bit: https://rust.godbolt.org/z/h47hTT

The general observation is that u128 _will be_ faster than checked u64 because it unrolls this loop. But it seems that this is currently only realized for nightly rustc... Stable rusts only manages to unroll for simple data shape like `&[u64]`. nightly even can handle a tuple like this. Our compiler is getting smarter. :)

Because checked u64 math introduces arbitrary abortion execution path at each addition step, it can't unroll.
So, I'll rewrite to use u128.

**Comment by ryoqun** [comment]
> I have still something to do (add test for overflow; benchmark checked u64 vs u128 for summing)

This is done. For benchmarking, I saw no noticeable difference and doing the cap calculation should be no harm because it's so small compared to hashing:

slower:

```
$ sudo chrt --rr 99 bash -c "RUST_LOG=debug  /tmp/solana-accounts-bench-slower --num_slots 4 --num_accounts 80000 > /tmp/result 2>&1" && cat /tmp/result | grep cap.*took
[2020-09-07T09:19:30.308469764Z DEBUG solana_runtime::accounts_db] sort took 558us hash took 24ms cap took 756us
[2020-09-07T09:19:30.366677331Z DEBUG solana_runtime::accounts_db] sort took 7ms hash took 19ms cap took 740us
[2020-09-07T09:19:30.417183194Z DEBUG solana_runtime::accounts_db] sort took 516us hash took 18ms cap took 786us
[2020-09-07T09:19:30.467481852Z DEBUG solana_runtime::accounts_db] sort took 627us hash took 18ms cap took 775us
[2020-09-07T09:19:30.518508803Z DEBUG solana_runtime::accounts_db] sort took 535us hash took 19ms cap took 764us
[2020-09-07T09:19:30.569019096Z DEBUG solana_runtime::accounts_db] sort took 558us hash took 18ms cap took 804us
[2020-09-07T09:19:30.620938190Z DEBUG solana_runtime::accounts_db] sort took 611us hash took 19ms cap took 760us
[2020-09-07T09:19:30.671447692Z DEBUG solana_runtime::accounts_db] sort took 569us hash took 19ms cap took 777us
[2020-09-07T09:19:30.722204665Z DEBUG solana_runtime::accounts_db] sort took 677us hash took 18ms cap took 769us
[2020-09-07T09:19:30.772587342Z DEBUG solana_runtime::accounts_db] sort took 602us hash took 18ms cap took 749us
[2020-09-07T09:19:30.823144280Z DEBUG solana_runtime::accounts_db] sort took 624us hash took 18ms cap took 719us
[2020-09-07T09:19:30.873514619Z DEBUG solana_runtime::accounts_db] sort took 573us hash took 18ms cap took 785us
[2020-09-07T09:19:30.924470309Z DEBUG solana_runtime::accounts_db] sort took 611us hash took 19ms cap took 777us
[2020-09-07T09:19:30.974693557Z DEBUG solana_runtime::accounts_db] sort took 584us hash took 18ms cap took 757us
[2020-09-07T09:19:31.025382499Z DEBUG solana_runtime::accounts_db] sort took 564us hash took 18ms cap took 751us
[2020-09-07T09:19:31.075546537Z DEBUG solana_runtime::accounts_db] sort took 599us hash took 18ms cap took 739us
[2020-09-07T09:19:31.125940844Z DEBUG solana_runtime::accounts_db] sort took 571us hash took 18ms cap took 766us
[2020-09-07T09:19:31.176431051Z DEBUG solana_runtime::accounts_db] sort took 639us hash took 18ms cap took 763us
[2020-09-07T09:19:31.227311366Z DEBUG solana_runtime::accounts_db] sort took 627us hash took 19ms cap took 748us
[2020-09-07T09:19:31.277557409Z DEBUG solana_runtime::accounts_db] sort took 640us hash took 19ms cap took 796us
```

faster:

```
$ sudo chrt --rr 99 bash -c "RUST_LOG=debug  /tmp/solana-accounts-bench-faster --num_slots 4 --num_accounts 80000 > /tmp/result 2>&1" && cat /tmp/result | grep cap.*took
[2020-09-07T09:19:42.578398620Z DEBUG solana_runtime::accounts_db] sort took 709us hash took 24ms cap took 759us
[2020-09-07T09:19:42.636548401Z DEBUG solana_runtime::accounts_db] sort took 7ms hash took 19ms cap took 706us
[2020-09-07T09:19:42.687420942Z DEBUG solana_runtime::accounts_db] sort took 610us hash took 18ms cap took 739us
[2020-09-07T09:19:42.738622939Z DEBUG solana_runtime::accounts_db] sort took 680us hash took 19ms cap took 711us
[2020-09-07T09:19:42.788517799Z DEBUG solana_runtime::accounts_db] sort took 645us hash took 18ms cap took 729us
[2020-09-07T09:19:42.839177491Z DEBUG solana_runtime::accounts_db] sort took 683us hash took 19ms cap took 722us
[2020-09-07T09:19:42.889007691Z DEBUG solana_runtime::accounts_db] sort took 610us hash took 18ms cap took 733us
[2020-09-07T09:19:42.940074238Z DEBUG solana_runtime::accounts_db] sort took 690us hash took 18ms cap took 712us
[2020-09-07T09:19:42.990138219Z DEBUG solana_runtime::accounts_db] sort took 659us hash took 19ms cap took 716us
[2020-09-07T09:19:43.040929840Z DEBUG solana_runtime::accounts_db] sort took 683us hash took 18ms cap took 737us
[2020-09-07T09:19:43.090966708Z DEBUG solana_runtime::accounts_db] sort took 636us hash took 18ms cap took 717us
[2020-09-07T09:19:43.141026689Z DEBUG solana_runtime::accounts_db] sort took 665us hash took 18ms cap took 753us
[2020-09-07T09:19:43.191746274Z DEBUG solana_runtime::accounts_db] sort took 612us hash took 18ms cap took 710us
[2020-09-07T09:19:43.242019966Z DEBUG solana_runtime::accounts_db] sort took 646us hash took 19ms cap took 716us
[2020-09-07T09:19:43.293032927Z DEBUG solana_runtime::accounts_db] sort took 691us hash took 19ms cap took 711us
[2020-09-07T09:19:43.343507060Z DEBUG solana_runtime::accounts_db] sort took 600us hash took 19ms cap took 739us
[2020-09-07T09:19:43.394155445Z DEBUG solana_runtime::accounts_db] sort took 715us hash took 18ms cap took 718us
[2020-09-07T09:19:43.444164476Z DEBUG solana_runtime::accounts_db] sort took 692us hash took 18ms cap took 737us
[2020-09-07T09:19:43.494308321Z DEBUG solana_runtime::accounts_db] sort took 593us hash took 18ms cap took 716us
[2020-09-07T09:19:43.544581596Z DEBUG solana_runtime::accounts_db] sort took 658us hash took 19ms cap took 730us
```

For completeness and pure curiosity, I've went into details and found that `u128` version is slightly (about `+1%`) faster actually:

https://docs.google.com/spreadsheets/d/1nsnNwkaTQ4pjpQKPwEhYcxhuhU0YCnvYm81Nmvf9zjA/edit#gid=0

**Comment by ryoqun** [comment]
it seems that #11996 has worked. testnet stopped to overflow.

@sakridge But, now this is blocked by #12112

**Comment by ryoqun** [comment]
Instead of #12112, https://github.com/solana-labs/solana/pull/12111 could unblock this pr more faster....

File: runtime/src/accounts.rs
Comment (sakridge) [inline runtime/src/accounts.rs:433]: Would like to move away from `load_by_program` with a `None` argument, this is kind of just an obfuscated `load_all` kind of function, right?

File: runtime/src/accounts_db.rs
Comment (sakridge) [reply runtime/src/accounts_db.rs:1693]: great. thanks

File: runtime/src/accounts_db.rs
Comment (sakridge) [inline runtime/src/accounts_db.rs:1700]: Just return 0?

Comment (sakridge) [reply runtime/src/accounts_db.rs:1700]: Or is it because anyone could transfer to sysvars? If that's the case, I think if we can reword the comment to say something like "specially retained accounts always have an initial 1 lamport balance, but could be modified by transfers which increase the balance but don't affect the capitalization"

File: runtime/src/accounts_db.rs
Comment (sakridge) [inline runtime/src/accounts_db.rs:1661]: any idea how much this adds to the time?

File: runtime/src/accounts_db.rs
Comment (sakridge) [inline runtime/src/accounts_db.rs:4556]: Should we have a test (probably in bank.rs) which does some rent and inflation and then checks the capitalization?

File: runtime/src/accounts_db.rs
Comment (sakridge) [reply runtime/src/accounts_db.rs:1661]: Ok, I see the other benchmarks have it at 700us for 80k signatures and seems to add about 5%. If so, that's not to bad.

File: runtime/src/accounts_db.rs
Comment (ryoqun) [reply runtime/src/accounts_db.rs:1700]: Comment clarified: 22c1f2c9ee90410e8ffef6960dcee15de28258a6

File: runtime/src/accounts_db.rs
Comment (ryoqun) [reply runtime/src/accounts_db.rs:4556]: Yeah, I've updated to verify capitalization in those relevant tests: 95cb371ecd6ba8dfb0527ee4e828081c1d0e937f

File: runtime/src/accounts.rs
Comment (ryoqun) [reply runtime/src/accounts.rs]: I neglected the confusing name... not anymore!: 21d6864ccf395ff86c510df2823277078d6dc20c

**Comment by ryoqun** [comment]
@sakridge Thanks for reviewing. I think I've addressed all your review comments.

File: runtime/src/accounts_db.rs
Comment (sakridge) [reply runtime/src/accounts_db.rs:4556]: awesome

**Review by sakridge — APPROVED** [review]
looks really good, thanks!

**Pull request merged**
